src/sqlfluff/testing/rules.py

```diff
@@ -141,4 +141,12 @@ def rules__test_helper(test_case):
         )
         # If a `fixed` value is provided then check it matches
         if test_case.fix_str:
-            assert res == test_case.fix_str
+            assert res == test_case.fix_str, "result of fix doesn't match the fix_str"
+            assert_rule_pass_in_sql(
+                test_case.rule,
+                test_case.fix_str,
+                configs=test_case.configs,
+            )
```

Comment:
If you change it to this, it will use the "noqa" comments when checking the fix_str, but not when making the original fixes. 😎 
```suggestion
            linted = Linter(config=test_case.configs).lint_string(test_case.fix_str)
            assert not linted.get_violations()
```

Reply:
I should probably take your word on it, but I can't see how that could work. The fixing seems to happen already before the code that you suggested to change, in `res = assert_rule_fail_in_sql(..`. I understood it so that noqa prevents any fixes to be done on the lines that have it.

But if I understood your idea correctly, that seems valid though: fix first so that noqa is ignored, then lint with noqa enabled.

But sorry, I still don't like it too much. I think it's not good to have a different meaning for noqa in tests than in runtime. I'm in favor of extending the testing framework.

I could imagine that addressing any complexity like this would be trivial if tests cases were written in python instead of using a custom yaml DSL. To me it seems that the only benefit of yaml is that it allows writing some things without surrounding quotes. But it makes it harder to run specific tests (especially in an IDE) and has limitations for what kind of things one is able to assert in the tests. But it would be nice to understand better what benefit the yaml approach brings, maybe I'm missing something.

Reply:
Personally, as a non-Python developer, I like the simplicity of the current test suite.

We do have Python tests outside of the fixtures directory for more complex tests (e.g. testing for two rules together when they’ve caused issues in the past, or testing that specific error message messages are returned for those rules that alter their error messages).

But those are the rare exceptions and the YAML syntax to quickly be able to paste in SQL (rather than some string value containing that SQL) and simple configuration makes test cases very easy to add, and approachable for new users - especially those without Python skills.

Reply:
I like the YAML tests because it's very easy to add them, very little "overhead" in terms of effort.

+1 for extending the testing framework to verify certain expected errors are present after fixing.

Reply:
We really should take a stab at this: https://github.com/sqlfluff/sqlfluff/issues/2180

And just realised that you raised this! 😁

Reply:
> I like the YAML tests because it's very easy to add them, very little "overhead" in terms of effort.

Ok, I suppose I can appreciate that as well. But still, based on my experience on developing these tests I would have to disagree, because I haven't been able to find a nice way to execute specific rule yaml tests in my IDE. Being able to run & debug regular python pytests is way more useful for my productivity than the possibility to avoid some minimal clutter in the test case syntax. + I like the power of python & the freedom it gives for eg. documenting the test cases better.

> +1 for extending the testing framework to verify certain expected errors are present after fixing.

This is nice to hear, but I'm afraid that might take longer to accomplish, so I'm proposing to just migrate the few cases that remain to python based tests for now.

Here's the first one, I hope you're ok with this approach at least as an initial solution:
https://github.com/sqlfluff/sqlfluff/pull/2681